packages/orchestrator/internal/server/main.go

```diff
@@ -59,22 +68,78 @@ func New() (*grpc.Server, error) {
 		return nil, fmt.Errorf("failed to create network pool: %w", err)
 	}
 
-	s := grpc.NewServer(
-		grpc.StatsHandler(e2bgrpc.NewStatsWrapper(otelgrpc.NewServerHandler())),
-		grpc.ChainUnaryInterceptor(
-			recovery.UnaryServerInterceptor(),
-		),
-	)
+	// BLOCK: initialize services
+	{
+		srv.dns = dns.New()
+		srv.grpc = grpc.NewServer(
+			grpc.StatsHandler(e2bgrpc.NewStatsWrapper(otelgrpc.NewServerHandler())),
+			grpc.ChainUnaryInterceptor(
+				recovery.UnaryServerInterceptor(),
+			),
+		)
+
+		srv.server = &server{
+			tracer:        otel.Tracer(ServiceName),
+			dns:           srv.dns,
+			sandboxes:     smap.New[*sandbox.Sandbox](),
+			networkPool:   networkPool,
+			templateCache: templateCache,
+		}
+	}
+
+	orchestrator.RegisterSandboxServiceServer(srv.grpc, srv.server)
+	grpc_health_v1.RegisterHealthServer(srv.grpc, health.NewServer())
+
+	return srv, nil
+}
+
+// Start launches
+func (srv *Service) Start(context.Context) error {
+	if srv.server == nil || srv.dns == nil || srv.grpc == nil {
+		return errors.New("orchestrator services are not initialized")
+	}
+
+	go func() {
+		log.Printf("Starting DNS server")
+		if err := srv.dns.Start("127.0.0.4", 53); err != nil {
+			log.Panic(fmt.Errorf("Failed running DNS server: %w", err))
+		}
+	}()
+
+	// the listener is closed by the shutdown operation
+	lis, err := net.Listen("tcp", fmt.Sprintf(":%d", srv.port))
+	if err != nil {
+		return fmt.Errorf("failed to listen on port %d: %w", srv.port, err)
+	}
+
+	log.Printf("starting server on port %d", srv.port)
+
+	go func() {
+		if err := srv.grpc.Serve(lis); err != nil {
+			log.Panic(fmt.Errorf("failed to serve: %w", err))
+		}
+	}()
+
+	srv.shutdown.op = func(ctx context.Context) error {
+		var errs []error
 
-	orchestrator.RegisterSandboxServiceServer(s, &server{
-		tracer:        otel.Tracer(ServiceName),
-		dns:           dnsServer,
-		sandboxes:     smap.New[*sandbox.Sandbox](),
-		networkPool:   networkPool,
-		templateCache: templateCache,
-	})
+		srv.grpc.GracefulStop()
 
-	grpc_health_v1.RegisterHealthServer(s, health.NewServer())
+		if err := srv.dns.Close(ctx); err != nil {
+			errs = append(errs, err)
+		}
+
+		if err := lis.Close(); err != nil {
+			errs = append(errs, err)
+		}
```

Comment:
I would probably switch these two, not that important. It's similar to API, you can probably serve some traffic if DNS is up, on the other side the server is in this point probably useless

Reply:
I think because we call graceful stop on line 126 that this doesn't matter, but I don't think  this order is super meaningful. 

There's the alternate argument that because all of the processes that requests would target are _children_ of the orchestrator, we should stop DNS as early as possible because we want traffic to cleanly halt and stopping DNS before we stop the workload means we we get a (potentially retriable) error rather than have a hanging request or return corrupt/partial data. 

I don't it particularly matters, and I've swapped the order.